apps/predictions/lib/predictions_pub_sub.ex

```diff
@@ -30,12 +30,23 @@ defmodule Predictions.PredictionsPubSub do
   end
 
   @spec subscribe(String.t()) :: [Prediction.t()]
-  @spec subscribe(String.t(), GenServer.server()) :: [Prediction.t()]
-  def subscribe(key, server \\ __MODULE__) do
-    StreamSupervisor.ensure_stream_is_started(key)
-    {registry_key, predictions} = GenServer.call(server, {:subscribe, key})
-    Registry.register(@subscribers, registry_key, key)
-    predictions
+  @spec subscribe(String.t(), GenServer.server()) :: [Prediction.t()] | {:error, term()}
+  def subscribe(topic, server \\ __MODULE__) do
+    with %StreamTopic{streams: streams, fetch_keys: fetch_keys} = stream_filter <-
+           StreamTopic.new(topic) do
+      Enum.each(streams, &StreamSupervisor.ensure_stream_is_started/1)
```

Comment:
suggestion: rather than having this module pull the contents out of `StreamTopic`, what about making this (and other functionality which works with data from this struct) a function in the `StreamTopic` module?